Alerting: Add a general screenshot service and alerting-specific image service.

joeblubaugh opened the pull request:
Supercedes #48449, since I can't work on a branch on that forked repo.

## What this PR does

This pull request adds a screenshot package for taking and uploading screenshots of Grafana dashboards. It supports taking screenshots of both dashboards and individual panels within a dashboard.

The screenshot package has the following services, most of which can be composed:

    BrowserScreenshotService (Takes screenshots with headless Chrome)
    CachableScreenshotService (Caches screenshots taken with another service such as BrowserScreenshotService)
    NoopScreenshotService (A no-op screenshot service for tests)
    SingleFlightScreenshotService (Prevents duplicate screenshots when taking screenshots of the same dashboard or panel in parallel)
    ScreenshotUnavailableService (A screenshot service that returns ErrScreenshotsUnavailable)
    UploadingScreenshotService (A screenshot service that uploads taken screenshots)

The screenshot package does not support wire dependency injection yet. ngalert constructs its own version of the service. See #49296 

This PR also adds an ImageScreenshotService to ngAlert. This is used to take screenshots and then store their location reference for use by alert instances and notifiers.

## What this PR does not do

This PR doesn't not take screenshots when alerting states change. For a follow-up.
This pull request does not update the notifiers to include screenshots in alert notifications. For a follow-up.

grobinson-grafana committed:
Screenshots in ngalert

This commit creates a screenshot service for taking and uploading
screenshots. A screenshot is taken in ngalert when an alert rule
changes from pending to firing.

grobinson-grafana committed:
Create ImageService and configuration in defaults.ini

joeblubaugh committed:
Finish the TODOs in prior PR.

Includes:
- Test for Image Database tables
- No error with ScreenshotsUnavailableService in state.Manager

joeblubaugh commented on pkg/services/ngalert/image/service.go:
```diff
@@ -0,0 +1,119 @@
+	if r.DashboardUID == nil {
+		return nil, ErrNoDashboard
+	}
+	if r.PanelID == nil {
```
Do we care if it's set but it's zero? We seem to assume it's nonzero in other places.

JohnnyQQQQ replied:
Zero will result in an invalid request, thus I think we should catch it here.

JohnnyQQQQ commented on pkg/services/ngalert/image/service.go:
```diff
@@ -0,0 +1,119 @@
+		}
+		s = screenshot.NewUploadingScreenshotService(metrics, s, u)
+	}
+	s = screenshot.NewRateLimitScreenshotService(s, cfg.UnifiedAlerting.Screenshots.MaxConcurrentScreenshots)
+	s = screenshot.NewSingleFlightScreenshotService(s)
+	s = screenshot.NewCachableScreenshotService(metrics, screenshotCacheTTL, s)
+	s = screenshot.NewObservableScreenshotService(metrics, s)
```
Nit(non-blocking): I think I would have done this differently, more go'ish, by doing something like

```golang
s := screenshot.NewService(withRateLimit(MaxConcurrentScreenshots), withCache(screenshotCacheTTL), withSingleFlight(),...)
````

joeblubaugh replied:
https://github.com/grafana/grafana/issues/new?assignees=&labels=type%3A+bug&template=1-bug_report.md

JohnnyQQQQ commented on pkg/services/ngalert/image/service.go:
```diff
@@ -0,0 +1,119 @@
+//go:generate mockgen -destination=mock.go -package=image github.com/grafana/grafana/pkg/services/ngalert/image ImageService
+type ImageService interface {
+	// NewImage returns a new image for the alert instance.
+	NewImage(ctx context.Context, r *ngmodels.AlertRule, labels data.Labels) (*store.Image, error)
```
That might be a silly question, but why do we need the `data.Labels`?

joeblubaugh replied:
For our glorious future where we have an image service that *can* render different views for different labels :-) 

I do think I'll remove it for now, though. We can always change the signature.

JohnnyQQQQ commented on pkg/services/ngalert/store/image.go:
```diff
@@ -0,0 +1,97 @@
+	ErrImageNotFound = errors.New("image not found")
+)
+
+type Image struct {
```
We should implement `TableName() string` on this struct and return the table name (`alert_image`), this makes sure that we don't need to define the table name below and can just use the struct.

```golang
if _, err := sess.Table("alert_image").Insert(img); err != nil {
    return fmt.Errorf("failed to insert screenshot: %w", err)
}
```
becomes

```golang
if _, err := sess.Insert(img); err != nil {
    return fmt.Errorf("failed to insert screenshot: %w", err)
}
```

joeblubaugh replied:
I like it.

joeblubaugh replied:
That Update path is driving me crazy, though. Every time I tried to just use that update call it would use the wrong key and violate a UNIQUE constraint.

JohnnyQQQQ reviewed (commented):
I did a first rough pass, looked good, left some comments/thoughts

JohnnyQQQQ commented on pkg/services/ngalert/store/image.go:
```diff
@@ -0,0 +1,97 @@
+				return fmt.Errorf("failed to insert screenshot: %w", err)
+			}
+		} else {
+			// TODO: Something is jacked up with XORM, and it's not properly setting the id field. Let's do it the hard way.
+			// This is what we want, but already-used id keys are getting overwritten
+			//if _, err := sess.Table("alert_image").Update(img); err != nil {
+			//	return fmt.Errorf("failed to update screenshot: %v", err)
+			//}
+
+			updateSQL := st.SQLStore.Dialect.UpsertSQL("alert_image",
+				[]string{"id"},
+				[]string{"id", "token", "path", "url", "created_at", "expires_at"})
+
+			_, err := sess.SQL(updateSQL, []interface{}{img.ID, img.Token, img.Path, img.URL, img.CreatedAt, img.ExpiresAt}...).Query()
+			if err != nil {
+				return fmt.Errorf("failed to update screenshot: %v", err)
+			}
+		}
```
This should do the trick, you have to pass in some condition otherwise it tries to insert 🤷 

```suggestion
			if _, err := sess.Table("alert_image").ID(img.ID).Update(img); err != nil {
				return fmt.Errorf("failed to update screenshot: %v", err)
			}
```

joeblubaugh committed:
Minor changes:

- remove labels from ImageService signature
- simpler xorm code
- fix a bug deleting expired alerts.

joeblubaugh committed:
Check for 0 panel id in addition to non-nil

alexweav commented on pkg/services/ngalert/store/image_test.go:
```diff
@@ -0,0 +1,129 @@
+func addToken(img *store.Image) *store.Image {
+	token, err := uuid.NewV4()
+	if err != nil {
+		panic("wat")
```
lol

alexweav commented on pkg/services/screenshot/screenshot.go:
```diff
@@ -0,0 +1,379 @@
+type BrowserScreenshotService struct {
+	ds                  dashboards.DashboardService
+	rs                  rendering.Service
+	screenshotDuration  prometheus.Histogram
+	screenshotFailures  prometheus.Counter
+	screenshotSuccesses prometheus.Counter
```
:thinking: is it typical to inject monitoring right next to dependencies? (Not a golang expert, so I'm not quite sure what is canonical here)

~~Curious as to your thoughts on adding another decorator to wrap the metrics? `MonitoredScreenshotService` or something to that effect?~~

I just realized `ObservableScreenshotService` decorator exists with nearly the exact same metrics. Why duplicate the counters?

joeblubaugh replied:
We shouldn't - I'll dedupe

joeblubaugh replied:
In general it seems fine to me - I'm not big on decorators in Go. As JP mentioned in his review, it's fairly common to inject a capability into a general implementation that normally has "do nothing" versions of those capabilities:

```go
type Opts struct {
  authFunc func(req) authResul
  rateLimiter rate.Limiter
}

alexweav commented on pkg/services/screenshot/screenshot.go:
```diff
@@ -0,0 +1,379 @@
+	opts = opts.SetDefaults()
+
+	// Compute the URL to screenshot.
+	renderPath := path.Join("d-solo", q.Result.Uid, q.Result.Slug)
```
Nit: Is there a constant we can re-use somewhere for this URL segment?

joeblubaugh replied:
Sure isn't. This is a broad problem in our entire API setup of the code - there are almost no path constants or programmatic construction anywhere. I don't know where I'd put the constants right now, tbh.

alexweav reviewed (approved):
LGTM! Just a couple nits and questions/optional feedback sorts of things.

JohnnyQQQQ reviewed (approved):
LGTM

joeblubaugh commented on pkg/services/ngalert/store/image.go:
```diff
@@ -0,0 +1,95 @@
+		// TODO: Is this a good idea?
+		img.ExpiresAt = TimeNow().Add(1 * time.Minute).UTC()
+		if img.ID == 0 { // xorm will generate fill this field on Insert.
+			token, err := uuid.NewV4()
```
I'm going to stick with sequential / token setup that's here, since it works with

joeblubaugh commented on pkg/services/ngalert/store/image.go:
```diff
@@ -0,0 +1,95 @@
+
+func (st DBstore) SaveImage(ctx context.Context, img *Image) error {
+	return st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		// TODO: Is this a good idea?
+		img.ExpiresAt = TimeNow().Add(1 * time.Minute).UTC()
```
Yes it is. I think if we're OK with the value I should just make it a constant.

joeblubaugh replied:
This actually isn't that. It's the expiration time that's set on a database row. `DeleteExpiredTokens` isn't actually called here yet, which is the only code that uses it. I didn't want to set it up, since a token row is tiny and the worst thing that's going to happen is a "Not Found" if you look up an old one.

 I'll expand the comment and link an issue for followup.

joeblubaugh committed:
Code review feedback: remove duplicate metrics, small cleanups

yuri-tceretian commented on pkg/services/ngalert/image/service.go:
```diff
@@ -0,0 +1,118 @@
+
+func (s *NoopImageService) NewImage(ctx context.Context, r *ngmodels.AlertRule) (*store.Image, error) {
+	return &store.Image{}, nil
+}
```
I wonder why it returns an empty struct instead of a nil.

joeblubaugh replied:
I'm up for deleting this service, personally. It caused me some testing headaches because it would fill empty structs in messages for any request. That's not a "Noop". It's why I ended up writing that `UnavailableImageService` later on. If I needed something for tests I'd rather just write a fake that stashes structs on save and does iterations for search.
